### DISCUSSION TITLE
Wrong arithmetic operation in spacing

### DISCUSSION THREAD

**Discussion by steida** [facebook/stylex#246] [Q&A]
I was surprised why Utopia spacing isn't working, and there is `-` instead of `+`.

This works:

```ts
export const spacing = stylex.defineVars({
  xxxs: `clamp(${MIN_SPACE.xxxs}px, calc(${INTERCEPT_SPACE.xxxs}px + ${
    Math.round(10000 * SLOPE_SPACE.xxxs) / 100
  }vw), ${MAX_SPACE.xxxs}px)`,
  xxs: `clamp(${MIN_SPACE.xxs}px, calc(${INTERCEPT_SPACE.xxs}px + ${
    Math.round(10000 * SLOPE_SPACE.xxs) / 100
  }vw), ${MAX_SPACE.xxs}px)`,
  xs: `clamp(${MIN_SPACE.xs}px, calc(${INTERCEPT_SPACE.xs}px + ${
    Math.round(10000 * SLOPE_SPACE.xs) / 100
  }vw), ${MAX_SPACE.xs}px)`,
  sm: `clamp(${MIN_SPACE.sm}px, calc(${INTERCEPT_SPACE.sm}px + ${
    Math.round(10000 * SLOPE_SPACE.sm) / 100
  }vw), ${MAX_SPACE.sm}px)`,
  md: `clamp(${MIN_SPACE.md}px, calc(${INTERCEPT_SPACE.md}px + ${
    Math.round(10000 * SLOPE_SPACE.md) / 100
  }vw), ${MAX_SPACE.md}px)`,
  lg: `clamp(${MIN_SPACE.lg}px, calc(${INTERCEPT_SPACE.lg}px + ${
    Math.round(10000 * SLOPE_SPACE.lg) / 100
  }vw), ${MAX_SPACE.lg}px)`,
  xl: `clamp(${MIN_SPACE.xl}px, calc(${INTERCEPT_SPACE.xl}px + ${
    Math.round(10000 * SLOPE_SPACE.xl) / 100
  }vw), ${MAX_SPACE.xl}px)`,
  xxl: `clamp(${MIN_SPACE.xxl}px, calc(${INTERCEPT_SPACE.xxl}px + ${
    Math.round(10000 * SLOPE_SPACE.xxl) / 100
  }vw), ${MAX_SPACE.xxl}px)`,
  xxxl: `clamp(${MIN_SPACE.xxxl}px, calc(${INTERCEPT_SPACE.xxxl}px + ${
    Math.round(10000 * SLOPE_SPACE.xxxl) / 100
  }vw), ${MAX_SPACE.xxxl}px)`,
  xxxxl: `clamp(${MIN_SPACE.xxxxl}px, calc(${INTERCEPT_SPACE.xxxxl}px + ${
    Math.round(10000 * SLOPE_SPACE.xxxxl) / 100
  }vw), ${MAX_SPACE.xxxxl}px)`,
});
```

Isn't it possible to refactor those computations into a lib? A "@stylexjs/typography" lib.

---

And MAX_BASE_SIZE isn't working too.

I have const MAX_BASE_SIZE = 18;

But p is 20px.

**Comment by necolas** [comment]
Please can you follow the issue template and try to provide actionable information in your reports?

> I was surprised why Utopia spacing isn't working

How is it not working? What are you expecting to see, and what are you actually seeing?

**Comment by nmn** [comment]
> Isn't it possible to refactor those computations into a lib? A "@stylexjs/typography" lib.

No because it's not possible to customize the numbers going in externally.

---

I'll look if I got the math wrong. I remember that at really small sizes things work a bit strangely, but I remember it all working correctly.

It might make sense to simply delete all the computation and replace it with the final output from Utopia directly. This was initially just a test to verify that evaluation during compilation was working correctly. We use the nextjs example as a bit of a testbed.

**Comment by steida** [comment]
> I was surprised why Utopia spacing isn't working

The spacing is the same, regardless of the browser viewport width.

I replaced `-` with `+`, and it adjusted size properly. 

During that time, I discovered another bug: the font size.

@necolas Sorry for the insufficient repro, I will add something.

@nmn Where did you get that math? It looks like https://utopia.fyi/ isn't open-sourced.

**Comment by nmn** [comment]
@steida I mostly came up with the math myself.

**Comment by steida** [comment]
@nmn OK, so it's probably out of the scope of StyleX. Should I close this issue?

**Reply by nmn** [reply]
It might be relevant if the example code is confusing to developers. I've converting to a discussion. I'll double-check my math and consider if we should remove it all in favour of a simple copy-paste.

**Comment by nmn** [comment]
First update: I validated the generated type variables in the code example against Utopia and it's almost exactly the same:

This is what is generated by StyleX with my math:

```css
--xbufh2v: clamp(0.58rem, calc(0.6rem + -0.09vw), 0.53rem);
--x1qaprzd: clamp(0.69rem, calc(0.69rem + 0.02vw), 0.7rem);
--x1owlcvw: clamp(0.83rem, calc(0.79rem + 0.19vw), 0.94rem);
--x18g1ocm: clamp(1rem, calc(0.91rem + 0.43vw), 1.25rem);
--x1wr3zdq: clamp(1.2rem, calc(1.04rem + 0.82vw), 1.67rem);
--x1m55loo: clamp(1.44rem, calc(1.17rem + 1.36vw), 2.22rem);
--x64pje0: clamp(1.73rem, calc(1.3rem + 2.14vw), 2.96rem);
--x1vlmfz8: clamp(2.07rem, calc(1.42rem + 3.27vw), 3.95rem);
--x8g1rd5: clamp(2.49rem, calc(1.53rem + 4.82vw), 5.26rem);
```

And this is what Utopia generates when given the same inputs:

```css
--step--3: clamp(0.5275rem, 0.5966rem + -0.0891vw, 0.5788rem);
--step--2: clamp(0.6944rem, 0.6911rem + 0.0163vw, 0.7038rem);
--step--1: clamp(0.8331rem, 0.7968rem + 0.1815vw, 0.9375rem);
--step-0: clamp(1rem, 0.913rem + 0.4348vw, 1.25rem);
--step-1: clamp(1.2rem, 1.0378rem + 0.8109vw, 1.6663rem);
--step-2: clamp(1.44rem, 1.1683rem + 1.3587vw, 2.2213rem);
--step-3: clamp(1.7281rem, 1.2994rem + 2.1435vw, 2.9606rem);
--step-4: clamp(2.0738rem, 1.4222rem + 3.2576vw, 3.9469rem);
--step-5: clamp(2.4881rem, 1.5238rem + 4.8217vw, 5.2606rem);
```

Second Update: I checked the spacing variables and you're right, I should be using a `+` and not a `-`.

Thanks for bringing this to my attention. Even though this is not a StyleX bug, it's nice to fix it!
